File: research/experiments/sweep/runs/opt-medium-cpp-opus45/src/test_finite_diff.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
#include "finite_diff.hpp"
#include <cmath>

using namespace opt;

// Test functions
static double sphere(const std::vector<double>& x) {
    double s = 0;
    for (auto v : x) s += v * v;
    return s;
}

static double rosenbrock(const std::vector<double>& x) {
    double a = 1.0 - x[0];
    double b = x[1] - x[0] * x[0];
    return a * a + 100.0 * b * b;
}

static std::vector<double> rosenbrockAnalyticGrad(const std::vector<double>& x) {
    double x0 = x[0], x1 = x[1];
    return {
        -2.0 * (1.0 - x0) + 200.0 * (x1 - x0 * x0) * (-2.0 * x0),
        200.0 * (x1 - x0 * x0)
    };
}

static double beale(const std::vector<double>& x) {
    double x1 = x[0], x2 = x[1];
    double t1 = 1.5 - x1 + x1 * x2;
    double t2 = 2.25 - x1 + x1 * x2 * x2;
    double t3 = 2.625 - x1 + x1 * x2 * x2 * x2;
    return t1 * t1 + t2 * t2 + t3 * t3;
}

static std::vector<double> bealeAnalyticGrad(const std::vector<double>& x) {
    double x1 = x[0], x2 = x[1];
    double t1 = 1.5 - x1 + x1 * x2;
    double t2 = 2.25 - x1 + x1 * x2 * x2;
    double t3 = 2.625 - x1 + x1 * x2 * x2 * x2;
    return {
        2.0 * t1 * (-1 + x2) + 2.0 * t2 * (-1 + x2 * x2) + 2.0 * t3 * (-1 + x2 * x2 * x2),
        2.0 * t1 * x1 + 2.0 * t2 * (2.0 * x1 * x2) + 2.0 * t3 * (3.0 * x1 * x2 * x2)
    };
}

TEST_CASE("forward diff: sphere at [3,4]") {
    auto g = forwardDiffGradient(sphere, {3, 4});
    CHECK(g[0] == doctest::Approx(6.0).epsilon(1e-5));
    CHECK(g[1] == doctest::Approx(8.0).epsilon(1e-5));
}

TEST_CASE("forward diff: sphere at [0,0]") {
    auto g = forwardDiffGradient(sphere, {0, 0});
    CHECK(std::abs(g[0]) < 1e-7);
    CHECK(std::abs(g[1]) < 1e-7);
}

TEST_CASE("forward diff: rosenbrock at [-1.2, 1.0]") {
    auto g = forwardDiffGradient(rosenbrock, {-1.2, 1.0});
    auto analytic = rosenbrockAnalyticGrad({-1.2, 1.0});
    CHECK(g[0] == doctest::Approx(analytic[0]).epsilon(1e-3));
    CHECK(g[1] == doctest::Approx(analytic[1]).epsilon(1e-3));
}

TEST_CASE("forward diff: beale at [1, 0.25] matches analytic") {
    auto numeric = forwardDiffGradient(beale, {1, 0.25});
    auto analytic = bealeAnalyticGrad({1, 0.25});
    CHECK(numeric[0] == doctest::Approx(analytic[0]).epsilon(1e-3));
    CHECK(numeric[1] == doctest::Approx(analytic[1]).epsilon(1e-3));
}

TEST_CASE("forward diff: does not mutate input") {
    std::vector<double> x = {3, 4};
    forwardDiffGradient(sphere, x);
    CHECK(x[0] == doctest::Approx(3));
    CHECK(x[1] == doctest::Approx(4));
}

TEST_CASE("central diff: sphere at [3,4] higher accuracy") {
    auto g = centralDiffGradient(sphere, {3, 4});
    CHECK(g[0] == doctest::Approx(6.0).epsilon(1e-8));
    CHECK(g[1] == doctest::Approx(8.0).epsilon(1e-8));
}

TEST_CASE("central diff: rosenbrock at [-1.2, 1.0] higher accuracy") {
    auto g = centralDiffGradient(rosenbrock, {-1.2, 1.0});
    auto analytic = rosenbrockAnalyticGrad({-1.2, 1.0});
    CHECK(g[0] == doctest::Approx(analytic[0]).epsilon(1e-5));
    CHECK(g[1] == doctest::Approx(analytic[1]).epsilon(1e-5));
}

TEST_CASE("central diff: beale at minimum [3, 0.5] is near-zero") {
    auto g = centralDiffGradient(beale, {3, 0.5});
    CHECK(std::abs(g[0]) < 1e-8);
    CHECK(std::abs(g[1]) < 1e-8);
}

TEST_CASE("central diff: does not mutate input") {
    std::vector<double> x = {3, 4};
    centralDiffGradient(sphere, x);
    CHECK(x[0] == doctest::Approx(3));
    CHECK(x[1] == doctest::Approx(4));
}

TEST_CASE("makeGradient: default is forward") {
    auto gf = makeGradient(sphere);
    auto g = gf({3, 4});
    auto gRef = forwardDiffGradient(sphere, {3, 4});
    CHECK(g[0] == doctest::Approx(gRef[0]));
    CHECK(g[1] == doctest::Approx(gRef[1]));
}

TEST_CASE("makeGradient: central method") {
    auto gf = makeGradient(sphere, "central");
    auto g = gf({3, 4});
    auto gRef = centralDiffGradient(sphere, {3, 4});
    CHECK(g[0] == doctest::Approx(gRef[0]));
    CHECK(g[1] == doctest::Approx(gRef[1]));
}

File: research/experiments/sweep/runs/opt-medium-python-sonnet45/src/test_finite_diff.py
"""Tests for finite-diff."""

import pytest
from finite_diff import forward_diff_gradient, central_diff_gradient, make_gradient


# Test functions
def sphere(x):
    """f(x) = sum(x_i^2)"""
    return sum(xi**2 for xi in x)


def sphere_grad(x):
    """Analytic gradient of sphere."""
    return [2 * xi for xi in x]


def rosenbrock(x):
    """Rosenbrock function."""
    return (1 - x[0])**2 + 100 * (x[1] - x[0]**2)**2


def rosenbrock_grad(x):
    """Analytic gradient of Rosenbrock."""
    return [
        -2 * (1 - x[0]) - 400 * x[0] * (x[1] - x[0]**2),
        200 * (x[1] - x[0]**2)
    ]


def beale(x):
    """Beale function."""
    return (
        (1.5 - x[0] + x[0] * x[1])**2 +
        (2.25 - x[0] + x[0] * x[1]**2)**2 +
        (2.625 - x[0] + x[0] * x[1]**3)**2
    )


def beale_grad(x):
    """Analytic gradient of Beale."""
    t1 = 1.5 - x[0] + x[0] * x[1]
    t2 = 2.25 - x[0] + x[0] * x[1]**2
    t3 = 2.625 - x[0] + x[0] * x[1]**3

    dx0 = (
        2 * t1 * (-1 + x[1]) +
        2 * t2 * (-1 + x[1]**2) +
        2 * t3 * (-1 + x[1]**3)
    )
    dx1 = (
        2 * t1 * x[0] +
        2 * t2 * x[0] * 2 * x[1] +
        2 * t3 * x[0] * 3 * x[1]**2
    )
    return [dx0, dx1]


def test_forward_diff_sphere_nonzero():
    """Test forward differences on sphere at [3, 4]."""
    x = [3.0, 4.0]
    grad_fd = forward_diff_gradient(sphere, x)
    grad_analytic = sphere_grad(x)

    assert grad_fd[0] == pytest.approx(grad_analytic[0], abs=1e-7)
    assert grad_fd[1] == pytest.approx(grad_analytic[1], abs=1e-7)


def test_forward_diff_sphere_zero():
    """Test forward differences on sphere at origin."""
    x = [0.0, 0.0]
    grad_fd = forward_diff_gradient(sphere, x)
    grad_analytic = sphere_grad(x)

    assert grad_fd[0] == pytest.approx(grad_analytic[0], abs=1e-7)
    assert grad_fd[1] == pytest.approx(grad_analytic[1], abs=1e-7)


def test_forward_diff_rosenbrock():
    """Test forward differences on Rosenbrock."""
    x = [-1.2, 1.0]
    grad_fd = forward_diff_gradient(rosenbrock, x)
    grad_analytic = rosenbrock_grad(x)

    assert grad_fd[0] == pytest.approx(grad_analytic[0], abs=1e-4)
    assert grad_fd[1] == pytest.approx(grad_analytic[1], abs=1e-4)


def test_forward_diff_beale():
    """Test forward differences on Beale."""
    x = [1.0, 1.0]
    grad_fd = forward_diff_gradient(beale, x)
    grad_analytic = beale_grad(x)

    assert grad_fd[0] == pytest.approx(grad_analytic[0], abs=1e-5)
    assert grad_fd[1] == pytest.approx(grad_analytic[1], abs=1e-5)


def test_central_diff_sphere_nonzero():
    """Test central differences on sphere at [3, 4]."""
    x = [3.0, 4.0]
    grad_cd = central_diff_gradient(sphere, x)
    grad_analytic = sphere_grad(x)

    assert grad_cd[0] == pytest.approx(grad_analytic[0], abs=1e-9)
    assert grad_cd[1] == pytest.approx(grad_analytic[1], abs=1e-9)


def test_central_diff_sphere_zero():
    """Test central differences on sphere at origin."""
    x = [0.0, 0.0]
    grad_cd = central_diff_gradient(sphere, x)
    grad_analytic = sphere_grad(x)

    assert grad_cd[0] == pytest.approx(grad_analytic[0], abs=1e-10)
    assert grad_cd[1] == pytest.approx(grad_analytic[1], abs=1e-10)


def test_central_diff_rosenbrock():
    """Test central differences on Rosenbrock."""
    x = [-1.2, 1.0]
    grad_cd = central_diff_gradient(rosenbrock, x)
    grad_analytic = rosenbrock_grad(x)

    assert grad_cd[0] == pytest.approx(grad_analytic[0], abs=1e-7)
    assert grad_cd[1] == pytest.approx(grad_analytic[1], abs=1e-7)


def test_central_diff_beale():
    """Test central differences on Beale."""
    x = [1.0, 1.0]
    grad_cd = central_diff_gradient(beale, x)
    grad_analytic = beale_grad(x)

    assert grad_cd[0] == pytest.approx(grad_analytic[0], abs=1e-8)
    assert grad_cd[1] == pytest.approx(grad_analytic[1], abs=1e-8)


def test_make_gradient_forward():
    """Test make_gradient with forward method."""
    grad_fn = make_gradient(sphere)
    x = [3.0, 4.0]
    grad = grad_fn(x)
    grad_analytic = sphere_grad(x)

    assert grad[0] == pytest.approx(grad_analytic[0], abs=1e-7)
    assert grad[1] == pytest.approx(grad_analytic[1], abs=1e-7)


def test_make_gradient_central():
    """Test make_gradient with central method."""
    grad_fn = make_gradient(sphere, "central")
    x = [3.0, 4.0]
    grad = grad_fn(x)
    grad_analytic = sphere_grad(x)

    assert grad[0] == pytest.approx(grad_analytic[0], abs=1e-9)
    assert grad[1] == pytest.approx(grad_analytic[1], abs=1e-9)
